### ISSUE TITLE
Viewport save_png does not save Alpha, and get_pixel make crash

### ISSUE THREAD

**Issue by danilw** [godotengine/godot#32894]
**Fix** https://github.com/godotengine/godot/issues/32894#issuecomment-543290503

**Godot version:**
3.1.1 stable

**OS/device including version:**
Linux editor and release template

**Issue description:**
1. Using save_png on Viewport texture does not save Alpha that Viewport texture has.
2. Using get_pixel on texture from Viewport make crash**(outside of editor)**

**Steps to reproduce:**
```
	var image = fbo.get_texture().get_data()
	image.save_png("1.png") #bug1 no alpha
	print(image.get_pixel(0,0)) #bug2 crash (outside of editor)
````

**Minimal reproduction project:**
https://danilw.github.io/godot-utils-and-other/BUG_Viewport_image.zip

**Comment by qarmin** [comment]
Can't reproduce crash in 3.2 alpha2, but it shows this error:
```
ERROR: get_pixel: Image must be locked with 'lock()' before using get_pixel().
   At: core/image.cpp:2387.
```

**Comment by bruvzg** [comment]
> but it shows this error

This is expected, image should be locked by using `.lock()` before any pixel manipulation.

> [Color get_pixel ( int x, int y ) const](https://docs.godotengine.org/en/latest/classes/class_image.html#class-image-method-get-pixel)
>
>Returns the color of the pixel at (x, y) if the image is locked. If the image is unlocked, it always returns a Color with the value (0, 0, 0, 1.0).

**Comment by danilw** [comment]
I see
thanks,it fix my crash

**Comment by Zylann** [comment]
@danilw is your viewport set to be `transparent`?

**Comment by bruvzg** [comment]
Check if `fbo.get_texture().has_alpha()` is true, saved PNG formats depends on texture format:

Related issue: https://github.com/godotengine/godot/issues/18882

**Comment by danilw** [comment]
it false I checked
why it false when it actualy has alpha
and print its color(after lock) show it has alpha

**Comment by bruvzg** [comment]
For some reason it's hardcoded to false for viewports: https://github.com/godotengine/godot/blob/master/scene/main/viewport.cpp#L120-L123

**Comment by Zylann** [comment]
> and print its color(after lock) show it has alpha

@danilw when you use `get_pixel()`, it returns a `Color`. The `Color` type always has R, G, B, A components, no matter the format of the image (they will just have a default value).

**Comment by danilw** [comment]
slow copy-pyxels does apply transparent from Viewport 
```
	var image = fbo.get_texture().get_data()
	var img = Image.new()
	img.create(1280, 720, false, Image.FORMAT_RGBA8)
	image.lock()
	img.lock()
	for a in range(1280):
		for b in range(720):
			img.set_pixel(a,b,image.get_pixel(a,b))
	image.unlock()
	img.unlock()
```

**Comment by danilw** [comment]
> > and print its color(after lock) show it has alpha
> 
> @danilw when you use `get_pixel()`, it returns a `Color`. The `Color` type always has R, G, B, A components, no matter the format of the image (they will just have a default value). What matters to check if an image has alpha is to check `image.get_format()`.

for 3D Viewport with HDR image is FORMAT_RGBAH = 15
for 2D Viewport with no HDR is FORMAT_RGBA8 = 5

so thank I font fix

**Set final Viewport to 2D and disable HDR** that will save PNG with alpha, and you *do not need hand-slow convert*

**Comment by danilw** [comment]
> @danilw then the problem maybe comes from `save_png()`? Viewports can be 16-bit, but Godot only saves 8-bit PNGs. So `save_png` converts to 8-bits, but perhaps it dropped alpha in the process?
> 
> try this instead:
> 
> ```gdscript
> image.convert(Image.FORMAT_RGBA8)
> image.save_png(...)
> ```

or this, this is more correct yes
thanks!

**Comment by Zylann** [comment]
btw, I had a look at `png_driver_common.cpp`, if the input format isn't 16-bit, it goes calling `detect_alpha()`, which semantically means that even if you want an alpha channel, but for some reason all alphas are opaque, Godot will not save your alpha channel (so if you load back that PNG in Godot, it won't have one).

Furthermore, `detect_alpha()` completely ignores 16-bit formats so it's probably why you don't get any when coming from a viewport.

**Comment by Byteron** [comment]
In 3.1 it worked for me but in 3.2 Alpha 1 and 2 my project crashes when calling `get_pixel`.

**Comment by KoBeWi** [comment]
Fixed in 4.0.

If you want transparency in 3.2, change the Viewport usage to 2D, because it works for some reason.

**Comment by mojoyup** [comment]
The above is not working for me in 3.4.4 with hdr false and converting to rgba8 (5). Ive tried with tree viewport, an attached viewport for 3d and also another one for 2d. Nothing works. It still captures the background of the environment color, which if I dont use that, I can see my desktop.....
